stewart_intro/simple_nn.py

```diff
@@ -0,0 +1,130 @@
+                cell.QT.values,
+                cell.SLI.values,
+                [cell.LHF.values],
+                [cell.SHF.values],
+                [cell.SOLIN.values]
+            ]
+        ) for cell in batch_values
+    ])
+    fqts_fslis = [
+        np.array(
+            [np.concatenate([cell.FQT.values, cell.FSLI.values])
+             for cell in batch_values])
+    ]
+    for i in range(1, n_time_steps + 1):
+        batch_values = [
+            data.sel(x=row.x, y=row.y, time=row.time + (dt * i))
+            for _, row in batch.iterrows()
+        ]
+        if i == n_time_steps:
+            target = np.array([
+                np.concatenate([cell.QT.values, cell.SLI.values])
+                for cell in batch_values
+            ])
+        if i != n_time_steps:
+            fqts_fslis.append(np.array([
+                np.concatenate([cell.FQT.values, cell.FSLI.values])
+                for cell in batch_values
+            ]))
+    return (
+        Variable(torch.from_numpy(x_data), requires_grad=True),
+        Variable(torch.from_numpy(target), requires_grad=True),
+        Variable(torch.from_numpy(np.array(fqts_fslis)), requires_grad=True)
+    )
+
+
+def initialize_weights(data):
+    w1 = Variable(
+        (torch.rand((len(data.z) * 2) + 3, n_hidden_nodes) - .5) / 1000,
+        requires_grad=True)
+    w2 = Variable((torch.rand(n_hidden_nodes, len(data.z) * 2) - .5) / 1000,
+                  requires_grad=True)
+    return w1, w2
+
+
+def predict(w1, w2, n_time_steps, x_data, fqts_fslis):
+    prediction = x_data[:, :-3]
+    for i in range(n_time_steps):
+        prediction = prediction + (dt * fqts_fslis[i])
+        to_predict_from = torch.transpose(torch.cat(
+            (
+                torch.transpose(prediction, 0, 1),
+                torch.transpose(x_data[:, -3:], 0, 1),
+            )
+        ), 0, 1)
+        layer_one_out = F.relu(to_predict_from.matmul(w1))
+        prediction = layer_one_out.matmul(w2)
+    return prediction
+
+
+def squared_loss(y_hat, y):
+    return (1 / 2) * (y - y_hat).pow(2)
+
+
+def compute_error(w1, w2, x_data, target, fqts_fslis):
+    if len(x_data.size()) == 1:
+        n = 1
+    else:
+        n = x_data.size()[0]
+    y_hat = predict(w1, w2, n_time_steps, x_data, fqts_fslis)
+    return ((1 / n) * squared_loss(
+        y_hat, target).sum())
+
+
+def train_model():
+    data = normalize_data()
+    batches = get_batches(data)
+    w1, w2 = initialize_weights(data)
+    n_batches = len(batches)
+    for epoch_num in range(n_epochs):
+        for idx, batch in enumerate(batches):
+            print(f'{idx} of {n_batches} for epoch {epoch_num}')
+            x_data, target, fqts_fslis = \
+                get_formatted_training_data_from_batch(data, batch)
+            err_est = compute_error(w1, w2, x_data, target, fqts_fslis)
+            print(float(err_est))
+            err_est.backward()
+            w1.data = w1.data - (training_rate * w1.grad.data)
```

Comment:
Props on implementing this fixed learning rate gradient descent manually. very cool. In the future, you will probably want to use `Adam` or some fancier optimizer. Torch has a bunch of implementations here: https://pytorch.org/docs/stable/optim.html#how-to-use-an-optimizer.

Reply:
Yes, I am sure the accelerated methods will converge a lot faster. I will use those going forward.

Reply:
yeah i am getting convergence, it just takes a while. What batch size do you find works well?

Reply:
maybe if you use a learning_rate of .001 it will be faster. Batch size shouldn't matter too much.

Reply:
Adam really does make a big difference TBH